### DISCUSSION TITLE
A/B partitioning implementation on MCU class devices necessary?  Can Mender be used as a transport leaving the firmware update to the device?

### DISCUSSION THREAD

**Discussion by maskimthedog** [joelguittet/mender-mcu-client#35] [Q&A]
A/B partitioning implementation on MCU class devices necessary?  Can Mender be used as a transport leaving the firmware update to the device?  The MCU on our gateways has it's own flash and we have SPI flash where we can store firmware packages received from the mender client prior to deployment to MCU flash.  However, the MCU flash does not have enough room to accommodate two partitions.  How to proceed?

**Comment by joelguittet** [comment]
Hello @maskimthedog 

Two points in your discussion:

- A/B partitioning is highly recommended because of two reasons: first you need to execute the firmware while you are downloading a copy that you will use to switch on. So you have an active partition and an update partition. Then because the update can fails for various reason it is highly recommended to have a rollback to a working copy in case of errors (network configuration issue leading to blocking communications, interface with other components in the device not working for a specific reason on a sub-set of the fleet, etc.)

- Using SPI flash: the mender-mcu-client is designed to fit this possibility as well. To achieve this, you should have your own implementation of the platform mender-ota functions and you should define `CONFIG_MENDER_PLATFORM_OTA_TYPE="generic/weak"` while building the library.
I have used this possibility on the [mender-stm32h745-zephyr-example](https://github.com/joelguittet/mender-stm32h745-zephyr-example) to have a custom mender-storage implementation specific to this evaluation board.
Basically you will need to write your own flash handlers so that the mender-mcu-client will be able to flash the external SPI memory. You will also need to implement the management of partitions in the bootloader to check if an update has been done, try the update, rollback in case of failure.

Please indicate me a lit bit more about your device so that I can guide you on this design, and I will be happy to do it! What is the bootloader? Which MCU are you using?

Joel

**Reply by maskimthedog** [reply]
Hi Joel,

We are doing a FreeRTOS based project using STM32F429 to demo Mender MCU OTA to a customer.  The customer wants to migrate from firmware upgrade via thumb drive in the field by a tech to OTA firmware upgrade.  They want to upgrade their systems in the field to support OTA and changing bootloader is not an option.  These systems are outdoors and see all sorts of weather.

[NUCLEO F429ZI](https://stackoverflow.com/questions/78044563/nucleo-f429zi-board-ethernet-over-usb-zperf-example-communication-issue)

Rob

**Reply by joelguittet** [reply]
Hi Rob,
Interesting challenge!
If the existing bootloader doesn't permit to flash the memory/do some partitions swap, then probably the best option is to add a second level bootloader that will be able to do so (installed the first time with the thumb drive together with the updated application mender client).
Would you like to send me a private email or message on [Linkedin](https://www.linkedin.com/in/jo%C3%ABl-guittet-59a41314a/) to discuss of the details? Maybe you don't want to share everything publicly here?
Joel

**Reply by joelguittet** [reply]
Rob,
By the way to your question "can Mender be used as a transport", the answer is definitively: YES. This is pure HTTP :-)
The end of the question is not clear for me.... "leaving the firmware update to the device" ...
Joel

**Reply by maskimthedog** [reply]
"leaving the firmware update to the device" ... I meant that the bootloader, as it exists today in the devices in the field, does unpack the firmware update package previously stored somewhere by the application from the thumb drive and programs it to flash.  Basically, the application using Mender as a transport and storing the firmware update package where the current bootloader can pick it up and program it to flash.

**Comment by maskimthedog** [comment]
I will provide you with my email address via LinkedIn.
